SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentReader.cpp

```diff
@@ -146,225 +146,286 @@ void ComponentReader::ApplyComponentUpdate(const Worker_ComponentId ComponentId,
+	Worker_ComponentId ComponentId = ApplySchemaObjectData.ComponentId;
+
 	TUniquePtr<FRepState>& RepState = Replicator->RepState;
 	TArray<FRepLayoutCmd>& Cmds = Replicator->RepLayout->Cmds;
 	TArray<FHandleToCmdIndex>& BaseHandleToCmdIndex = Replicator->RepLayout->BaseHandleToCmdIndex;
 	TArray<FRepParentCmd>& Parents = Replicator->RepLayout->Parents;
 
-	bool bIsAuthServer = Channel.IsAuthoritativeServer();
-	bool bAutonomousProxy = Channel.IsClientAutonomousProxy();
-	bool bIsClient = NetDriver->GetNetMode() == NM_Client;
-	bool bEventTracerEnabled = EventTracer != nullptr;
+	const Worker_EntityId EntityId = Channel.GetEntityId();
+	const bool bEventTracerEnabled = EventTracer != nullptr;
 
-	FSpatialConditionMapFilter ConditionMap(&Channel, bIsClient);
-
-	TArray<GDK_PROPERTY(Property)*> RepNotifies;
-	TMap<GDK_PROPERTY(Property)*, FSpatialGDKSpanId> PropertySpanIds;
+	const bool bIsAuthServer = Channel.IsAuthoritativeServer();
+	const bool bAutonomousProxy = Channel.IsClientAutonomousProxy();
+	const bool bIsClient = NetDriver->GetNetMode() == NM_Client;
+	const FSpatialConditionMapFilter ConditionMap(&Channel, bIsClient);
 
+	for (uint32 FieldId : ApplySchemaObjectData.UpdatedIds)
 	{
-		// Scoped to exclude OnRep callbacks which are already tracked per OnRep function
-		SCOPE_CYCLE_COUNTER(STAT_ReaderApplyPropertyUpdates);
-
-		Worker_EntityId EntityId = Channel.GetEntityId();
-		TArray<FSpatialGDKSpanId> CauseSpanIds;
-		if (bEventTracerEnabled)
+		// FieldId is the same as rep handle
+		if (FieldId == 0 || (int)FieldId - 1 >= BaseHandleToCmdIndex.Num())
 		{
-			CauseSpanIds = EventTracer->GetAndConsumeSpansForComponent(EntityComponentId(EntityId, ComponentId));
+			UE_LOG(LogSpatialComponentReader, Error,
+				   TEXT("ApplySchemaObject: Encountered an invalid field Id while applying schema. Object: %s, Field: %d, Entity: "
+						"%lld, Component: %d"),
+				   *Object.GetPathName(), FieldId, Channel.GetEntityId(), ComponentId);
+			continue;
 		}
 
-		for (uint32 FieldId : UpdatedIds)
+		int32 CmdIndex = BaseHandleToCmdIndex[FieldId - 1].CmdIndex;
+		const FRepLayoutCmd& Cmd = Cmds[CmdIndex];
+		const FRepParentCmd& Parent = Parents[Cmd.ParentIndex];
+		const int32 ShadowOffset = Cmd.ShadowOffset;
+
+		// Special Case properties are properties that are used in building the FSpatialConditionMapFilter,
+		// and the client must read them before rebuilding the condition map for all other properties.
+		const bool bProcessingSpecialCases = ApplySchemaObjectData.ProcessFieldType == EProcessFieldType::OnlySpecialCase;
+		if (ApplySchemaObjectData.ProcessFieldType != EProcessFieldType::All)
 		{
-			// FieldId is the same as rep handle
-			if (FieldId == 0 || (int)FieldId - 1 >= BaseHandleToCmdIndex.Num())
+			auto ShouldConsiderProperty = [&Cmd, &bProcessingSpecialCases]() -> bool {
+				if (bProcessingSpecialCases)
+				{
+					return UNLIKELY((int32)AActor::ENetFields_Private::RemoteRole == Cmd.ParentIndex)
```

Comment:
hmm I think generally UNLIKELY is used for immediate branch prediction, so may be better to remove these and just put it around the `if (!bShouldConsiderProperty)`

Reply:
Sure, I was just going based off of what is done slightly further down inside GetSwappedCmd, which uses UNLIKELY when checking against Cmd.ParentIndex as well.